yt_geotiff/data_structures.py

```diff
@@ -185,12 +186,53 @@ def _count_grids(self):
         self.num_grids = 1
 
 
+class JPEG2000Hierarchy(GeoTiffHierarchy):
+    grid = GeoTiffGrid   
+
+    def _detect_output_fields(self):
+        # check data dir of the given jp2 file and grab all similarly named files.
+
+        # List of band files in s2 directory
+        s2_band_file_list = [os.path.basename(x) \
+         for x in glob.glob(self.ds.directory+'/*_***_***.jp2')]
+
+        # Follow example for GeoTiffHierarchy to populate the field list.
+        self.field_list = []
+        self.ds.field_units = self.ds.field_units or {}
+
+        # Filename dictionary
+        self.ds._field_filename = {}
+
+        # extract s2 band name from file name
+        def get_band_name(band_file_list):
+            band_file = band_file_list.split("_")
+            return (band_file)
+
+        band_names = [list(b) for b in zip(*map(get_band_name, s2_band_file_list))][2]
+
+        # Attribute with resolution of original loaded image
+        #self.ds._load_resolution = self.ds.resolution.d[0]
+
+        for _i in range(1, len(s2_band_file_list) + 1):
+            with rasterio.open(self.ds.directory+"/"+s2_band_file_list[_i-1], "r") as f:
+                group = 'bands'
+                field_name = (group, band_names[_i-1] +'_'+str(round(f.res[0])))
+                self.field_list.append(field_name)
+                self.ds.field_units[field_name] = ""
+                self.ds._field_filename.update({(band_names[_i-1] +'_'+str(round(f.res[0]))):\
+                 {'filename': (self.ds.directory+'/'+s2_band_file_list[_i-1]), 'resolution': f.res[0]}})       
```

Comment:
The value for the `filename` entry is the same as the filename you open with rasterio. I would suggest making a temporary variable that you use above and here to avoid having to do the same thing twice.

Similarly, one line above, you can use `field_name[1]` in place of `(band_names[_i-1] +'_'+str(round(f.res[0])))`.